datalad_next/tree.py

```diff
@@ -0,0 +1,1130 @@
+    ]
+
+    @staticmethod
+    @datasetmethod(name='tree')
+    @eval_results
+    def __call__(
+            path='.',
+            *,
+            depth=None,
+            dataset_depth=None,
+            include_files=False,
+            include_hidden=False):
+
+        if dataset_depth is not None:
+            # special tree defined by subdataset nesting depth
+            tree_cls = DatasetTree
+            dataset_tree_args = {"max_dataset_depth": dataset_depth}
+        else:
+            # simple tree defined by directory depth
+            tree_cls = Tree
+            dataset_tree_args = {}
+
+        tree = tree_cls(
+            Path(path),
+            max_depth=depth,
+            exclude_node_func=build_excluded_node_func(
+                include_hidden=include_hidden, include_files=include_files),
+            **dataset_tree_args
+        )
+
+        for node in tree.generate_nodes():
+            # yield one node at a time to improve UX / perceived speed
+            res_dict = {
+                "action": "tree",
+                "path": str(node.path),
+                "type": node.TYPE,
+                "depth": node.depth,
+                "exhausted_levels": list(tree.exhausted_levels),
+                "count": {
+                    "datasets": tree.node_count["DatasetNode"],
+                    "directories": tree.node_count["DirectoryNode"],
+                    **({"files": tree.node_count["FileNode"]}
+                       if include_files else {})
+                },
+            }
+            if node.TYPE == "dataset":
+                res_dict.update({
+                    "dataset_depth": node.ds_depth,
+                    "dataset_abs_depth": node.ds_absolute_depth,
+                    "dataset_is_installed": node.is_installed
+                })
+
+            if node.is_symlink():
+                # TODO: should we inform if the symlink is recursive (as per
+                #  `tree.is_recursive_symlink()`) although not broken? The
+                #  UNIX 'tree' command shows the message '[recursive,
+                #  not followed]' next to the path. Not sure if this is
+                #  interesting at all or more confusing.
+                res_dict["symlink_target"] = node.get_symlink_target()
+                res_dict["is_broken_symlink"] = node.is_broken_symlink()
+
+            if node.exception is not None:
+                # mimic error message of unix 'tree' command for
+                # permission denied error, otherwise use exception short
+                # message
+                message = "error opening dir" \
+                    if node.exception.name == "PermissionError" \
+                    else node.exception.message
+
+                yield get_status_dict(
+                    status="error",
+                    message=message,
+                    exception=node.exception,
+                    **res_dict
+                )
```

Comment:
RE: error handling: I added an attribute `exception` on the `_TreeNode` to store any exception (wrapped in a `CapturedException`) that may have occurred when creating the object, e.g. `PermissionError` when trying to access a directory. The goal is to display the error and continue yielding the rest of the tree without crashing.

Reply:
Preserving the exception via `CapturedException` is exactly how it should be done. We have some convention how to label them in a result record. I will cross-reference that, when going through the code again.

Reply:
When testing 26e2f178704634a1ae1c2f58d02712940879e76e I found a condition that still causes `tree` to crash.

```sh
$ tree cups/ssl
cups/ssl
├── server.crt -> /etc/ssl/certs/ssl-cert-snakeoil.pem
└── server.key -> /etc/ssl/private/ssl-cert-snakeoil.key

0 directories, 2 files
```

vs.

```sh
$ datalad tree cups/ssl
/<abspath cut>/cups/ssl
[ERROR  ] [Errno 13] Permission denied: '/<abspath cut>/cups/ssl/server.key' (PermissionError) 
```

with

```sh
% ls -la cups/ssl/server.key
lrwxrwxrwx 1 mih mih 38 Dec 25  2012 cups/ssl/server.key -> /etc/ssl/private/ssl-cert-snakeoil.key
```

where the target exists, but `/etc/ssl/private` is not readable:

```sh
% ls -l /etc/ssl
total 36
drwxr-xr-x 3 root root     16384 Jun 27 11:16 certs/
-rw-r--r-- 1 root root     12441 Jun  4 15:25 openssl.cnf
drwx--x--- 2 root ssl-cert  4096 Jun  5 10:27 private/
```

Sidenote: When the `path` argument given to `tree` is relative, the command renderer would start with an absolute path as the tree root, while the POSIX `tree` would output a relative path.

Reply:
> When testing https://github.com/datalad/datalad-next/commit/26e2f178704634a1ae1c2f58d02712940879e76e I found a condition that still causes tree to crash.

Good catch! I added this case to the test for broken symlinks. (I assume that a symlink to a file in an inaccessible directory is considered broken, since we cannot verify the existence of its target. `tree` also displays it as broken.)

The test passes now, but checks for permission errors are still not robust enough. Every `is_dir()` call is vulnerable. I will try to centralize error handling in the node constructor so I can set the `exception` property right away. This means instantiating nodes immediately, even if they will not be yielded in the end (hoping it does not impose too much of a performance cost). This may make other components cleaner as well, which currently operate only on paths and need to duplicate some of the node-creation logic.

Reply:
> Sidenote: When the path argument given to tree is relative, the command renderer would start with an absolute path as the tree root, while the POSIX tree would output a relative path.

This is a deliberate deviation from `tree` (see tests `test_root_path_is_normalized` and `test_no_difference_if_root_path_absolute_or_relative`). I always want to see the root printed as full path, without having to prefix it with `$PWD`. Other than for consistency with `tree` (and `ls`, `find`, ...), I don't see a reason to use relative paths. Maybe they help with readability if the root path is very long. This is my personal preference though. What's your take?

Reply:
That works for me. Thanks for the explanation!